cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+    else:
+        returns = torch.zeros_like(rewards)
+        for t in reversed(range(total_steps)):
+            if t == total_steps - 1:
+                nextnonterminal = 1.0 - next_done
+                next_return = next_value
+            else:
+                nextnonterminal = 1.0 - dones[t + 1]
+                next_return = returns[t + 1]
+            returns[t] = rewards[t] + gamma * nextnonterminal * next_return
+        advantages = returns - values
+    return advantages, returns
+
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+class Agent(nn.Module):
+    def __init__(self, envs):
+        super().__init__()
+        self.network = nn.Sequential(
+            layer_init(nn.Conv2d(4, 32, 8, stride=4)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(32, 64, 4, stride=2)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(64, 64, 3, stride=1)),
+            nn.ReLU(),
+            nn.Flatten(),
+            layer_init(nn.Linear(64 * 7 * 7, 512)),
+            nn.ReLU(),
+        )
+        self.actor = layer_init(nn.Linear(512, envs.single_action_space.n), std=0.01)
+        self.critic = layer_init(nn.Linear(512, 1), std=1)
+
+    def forward(self, x):
+        """Run policy and value functions."""
+        hidden = self.network(x)
+        logits = self.actor(hidden)
+        value = self.critic(hidden).squeeze(-1)
+        return logits, value
+
+    @staticmethod
+    def sample_action(logits):
+        probs = Categorical(logits=logits)
+        action = probs.sample()
+        return action, probs.log_prob(action)
+
+    @staticmethod
+    def action_log_prob(logits, action):
+        probs = Categorical(logits=logits)
+        return probs.log_prob(action), probs.entropy()
+
+    @staticmethod
+    def action_kl_divergence(old_logits, new_logits):
+        old_probs = Categorical(logits=old_logits)
+        new_probs = Categorical(logits=new_logits)
+        return kl_divergence(old_probs, new_probs).mean()
+
+
+def main():
+    args = parse_args()
+    run_name = f"{args.env_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # Fix multiplication precision on Ampere devices (30xx) prior to PyTorch 1.12
+    # https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere
+    torch.backends.cuda.matmul.allow_tf32 = False
```

Comment:
Could you do some preliminary experiments to check if this makes a difference? It's important that we understand the effect of the techniques added.

Reply:
unfortunately I don't have access to Ampere devices, so maybe I'll revert this as well

Reply:
I have ampere devices - once things are finalized I will be happy to give it a try.